core/integration/internal/engine.go

```diff
@@ -0,0 +1,400 @@
+package internal
```

Comment:
This may be the topic of our meeting soon, but either way one possibility for getting rid of this duplication is to push the dev engine image to the local registry we setup as part of the test suite in `internal/mage`. Then if tests like these need to reference it they can just use that base image rather than re-running all the instructions (and optionally layer whatever customizations they might need on top).

Pushing that image to the registry service is easier said than done due to the fact that we don't have host proxying for services (yet), but you can get it to work by doing a tarball export, getting that file into an exec and pushing it from there to the registry service.

I had to do something very similar recently in a different context, the go-containerregistry library was very useful: https://github.com/google/go-containerregistry

Here's an extract from that to give a rough idea (`tarball`, `name` and `remote` are packages from go-containerregistry):
```go
	_, err = devEngine.Export(ctx, "/tmp/engine.tar")
	if err != nil {
		return err
	}
	img, err := tarball.ImageFromPath("/tmp/engine.tar", nil)
	if err != nil {
		return err
	}
	ref, err := name.ParseReference(fmt.Sprintf("%s/engine:dev", registryHost))
	if err != nil {
		return err
	}
	if err := remote.Write(ref, img); err != nil {
		return err
	}
```

Reply:
Ah no worries, I just thought it might be slightly quicker to implement, it's pretty much the exact same idea as the registry, so either approach is all good with me 👍